wordpress-mobile/WordPress-iOS
# Pages List performance improvements

Fixes #21813. Also, in general, this PR makes it's possible to load a site with a large number of pages (300+).

## Issue

Currently Pages List performs very poorly on a site with a large number of pages. A pretty extreme case is #21813. The app may become unresponsive in 5 minutes or so, when about 6000 pages are loaded.

If you have a test site with 300 pages, you can open that site in the app and go to Pages screen and the app will become unresponsive very quickly.

If you don't have a test site with that many pages, you can use FG as the test site, with a little bit of code change: adding another "and" condition to [this if statement](https://github.com/wordpress-mobile/WordPress-iOS/blob/23.7.0.0/WordPress/Classes/Services/PostService.m#L135): `&& loadedPosts.count < 500`. The number 500 is the max of pages you want to the app to fetch. You can tweak this number to test the Pages List's performance against certain number of pages.

## Changes

There are a few issues involved:
1. [Pages are loaded in memory](https://github.com/wordpress-mobile/WordPress-iOS/blob/23.7.0.0/WordPress/Classes/Services/PostService.m#L135) and populated into UI after all pages are fetched from API. This gives a false sense of Pages List stuck at "Fetching pages...".
2. Pages are saved using the main context. Meaning all the Core Data operations, inserting, updating, deleting, happen on the main thread.
3. Pages are displayed in hierarchical order. But the current algorithm doesn't perform well. See https://github.com/wordpress-mobile/WordPress-iOS/pull/22056.

The first two issues are addressed by [the `PostRepository.fetchAllPages` function](https://github.com/wordpress-mobile/WordPress-iOS/blob/23.7.0.0/WordPress/Classes/Services/PostRepository.swift#L393). And the last issues is addressed by the linked PR #22056. This PR basically uses those new implementations to swap out the current problematic ones.

That means, this PR should not bring any UI changes to the app, it only addresses performance issues.

> [!NOTE]
> It'd be much easier to review this PR commit by commit.

## Further considerations

In other discussions, the general consensus is it's bad to load all pages automatically, instead of page by page as user scrolls through the list (like Posts List). @kean mentioned we could change Pages List to match WP.com, where the pages are displayed in a hierarchical order if there are certain number in total (say 600), but they are displayed page by page (like Posts List) if the total amount exceed this threshold.

I agree with that idea, but did not implement it in this PR, because I don't want this PR to get too large. I'm not sure if I'd have time to implement it though, because I have other project lined up (dropping Alamofire from WordPressKit).

## Test Instructions

Verify #21813 is fixed. Use FG to test Pages List performance.

This PR also adds a cancellation mechanism to the Pages List, where the Pages List stops loading more pages when user taps the "Back" button. You can verify this behaviour by using Xcode's "Network Activity Report" in the Debug panel: You should see periodic spikes in the Network Activity when the app are on FG's pages list and see no more spikes when exits the pages list screen.

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Test using FG.

3. What automated tests I added (or what prevented me from doing so)
None.

PR submission checklist:


UI Changes testing checklist: N/A
